platform_disk_api/project_namespace_migration_job.py

```diff
@@ -0,0 +1,356 @@
+
+logger = logging.getLogger(__name__)
+
+
+CURRENT_NAMESPACE = "platform-jobs"
+
+
+class DiskMigrationError(Exception):
+    ...
+
+
+class PvcInUseError(DiskMigrationError):
+    ...
+
+
+async def migration_loop(
+    config: JobMigrateProjectNamespaceConfig,
+    disk_ids_filter: Optional[set[str]] = None,
+) -> None:
+    """
+    A main loop which iterates over all disks and tries to migrate them to a
+    project-based namespaces.
+
+    :param config: ...
+    :param disk_ids_filter: optional filtering by disk IDs
+    """
+    async with create_kube_client(config.kube) as kube_client:
+        # get all disk namings
+        disk_namings = await kube_client.list_disk_namings()
+        pvc_to_disk_naming = {
+            dn.disk_id: dn for dn in disk_namings
+        }
+
+        # get all the PVCs which are apolo disks
+        pvc_url = URL(
+            kube_client._generate_pvc_url(namespace=CURRENT_NAMESPACE)
+        ).with_query(
+            labelSelector=f"{DISK_API_MARK_LABEL}=true"
+        )
+        all_pvc = await kube_client.get(pvc_url)
+
+        for pvc in all_pvc['items']:
+            pvc_name = pvc["metadata"]["name"]
+            if disk_ids_filter and pvc_name not in disk_ids_filter:
+                continue
+
+            # PV name (if bound).
+            # sometimes users can create disks, but not mount them;
+            # in such a case - PVC won't have a respective PV yet
+            pv_name = pvc["spec"].get("volumeName")
+
+            disk_naming = pvc_to_disk_naming.get(pvc_name)
+            if not disk_naming:
+                logger.error(
+                    "PVC does not have a disk naming. Skipping",
+                    extra={
+                        "pvc_name": pvc_name
+                    }
+                )
+                continue
+
+            try:
+                await migrate_disk(
+                    kube_client,
+                    pvc_name,
+                    pvc=pvc,
+                    disk_naming=disk_naming,
+                    pv_name=pv_name,
+                )
+            except DiskMigrationError:
+                logger.warning(
+                    "unable to migrate PVC",
+                    extra={
+                        "pv_name": pv_name,
+                        "pvc_name": pvc_name,
+                    }
+                )
+
+
+async def migrate_disk(
+        client: KubeClient,
+        pvc_name: str,
+        pvc: dict[str, Any],
+        disk_naming: DiskNaming,
+        pv_name: Optional[str] = None,
+) -> None:
+    """
+    an entry-point for a single disk migration.
+
+    `pv_name` might be missing, in a cases when disk was created,
+    but was never mounted to any job/app/etc.
+    """
+    logger.info("migrating disk: %s", pvc_name)
+
+    current_meta = pvc["metadata"]
+    org_name = current_meta["labels"].get(DISK_API_ORG_LABEL) or normalize_name(NO_ORG)
+    project_name = current_meta["labels"][PROJECT_LABEL]
```

Comment:
if for some reason there is no project label let's use user name as project name

Reply:
done